File: 3SumClosest.cpp
// 16 Array medium

#include <vector>
class Solution {
  public:
    int threeSumClosest(vector<int>& nums, int target) {
        if ( nums.size() < 3 ) {
            return 0;
        }
        sort( nums.begin(), nums.end() );
        int ret = nums[0] + nums[1] + nums[2];
        for ( size_t i = 0; i < nums.size() - 2; i++ ) {
            int front = i + 1;
            int end = nums.size() - 1;
            // 寻找有序数组两数和最接近t，采用两边收缩遍历法
            while ( end > front ) {
                int cur = nums[i] + nums[ front ] + nums[ end ];
                if(cur == target) return target;
                if ( abs(target-cur) < abs(target-ret) ) {
                    ret = cur;
                }
                if(cur > target){
                    end--;
                } else {
                    front++;
                }
            }
            // 跳过后面相同的nums[i]，防止重复
            while ( i + 1 < nums.size() && nums[ i + 1 ] == nums[ i ] ) {
                i++;
            }
        }
        return ret;
    }
};

File: 3sumClosest.js
// 16 Array medium

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
  if (nums.length < 3) return 0
  nums.sort((a,b) => a-b)
  let ret = nums[0] + nums[1] + nums[2]
  for (let i = 0; i < nums.length - 2; i++) {
    let left = i+1, right = nums.length-1
    while (left < right) {
      const sum = nums[i] + nums[left] + nums[right]
      if (sum === target) return target
      if (Math.abs(target-sum) < Math.abs(target-ret)) {
        ret = sum
      }
      if (sum < target) left++
      else right--
    }
  }
  return ret
}

console.log(threeSumClosest([-1, 2, 1, -4], 1))